language/vm/src/binary_format.rs

```diff
@@ -0,0 +1,362 @@
+}
+
+//
+// Specifications
+//
+// Specifications point to a generic handle and its instantiation.
+// The instantiation can be partial.
+// So for types a StructSpec is either S<T, W>, S<u8, bool>, S<T, u8>
+
+// A complete or partial instantiation of a generic struct
+pub struct StructSpec {
+    pub handle: StructHandleIndex,
+    pub instantiation: SignatureIndex,
+}
+
+// A complete or partial instantiation of a function
+pub struct FunctionSpec {
+    pub handle: FunctionHandleIndex,
+    pub instantiation: SignatureIndex,
+}
+
+// A complete or partial instantiation of a field (or the type of it).
+// A FieldSpec points to a generic FieldHandle and the owner type instantiation.
+pub struct FieldSpec {
+    pub field_spec: FieldHandleIndex,
+    pub instantiation: SignatureIndex,
+}
+
+
+//
+// Definitions
+//
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct StructDefinition {
+    pub struct_handle: StructHandleIndex,
+    pub layout: StructLayout,
+}
+
+// Fields are inlined and each struct carries the fields it owns.
+// The instructions that use fields ([imm]borrow_field) now have to specify
+// the type that owns the field and the field index.
+// For example, if you are referring to a field of `Cup<T>` instantiated
+// let c: cup<int> = ...;
+// let tea_type = &c.f // f typed T
+// here we want to put in the instruction borrow_field (Cup<int>, 1) // assuming f is field 1
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum StructLayout {
+    Native,
+    Declared(Vec<FieldDefinition>),
+}
+
+// no reason to carry the owner in the FieldDefinition now, as it is implicit
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FieldDefinition {
+    pub name: IdentifierIndex,
+    pub signature: Type,
+}
+
+// we want a better way to define native functions. That encoding is in `Code` which can
+// be native or move bytecoces
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct FunctionDefinition {
+    pub function: FunctionHandleIndex,
+    pub flags: u8, // this is just the `public` flag now
+    pub acquires_global_resources: Vec<StructDefinitionIndex>,
+    pub code: Code,
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub enum Code {
+    Native,
+    Declared(CodeUnit),
+}
+
+pub struct CodeUnit {
+    // ... max_stack_size is gone...
+
+    // locals carry only the locals signature without the arguments.
+    // However there is a single instruction that access locals uniformly,
+    // that is, as if there was a single vector of arguments + locals.
+    // The compiler (and generally clients) are supposed to do the math properly
+    pub locals: SignatureIndex,
+    pub code: Vec<Bytecode>,
+}
+
+//
+// Bytecodes
+//
+
+// The main issue to digest here is that every bytecode that deals with generic is now
+// split in 2 versions. The non generic and the generic. That has some advantages
+// we have talked about and we may touch on more
+pub enum Bytecode {
+    Pop,
+    Ret,
+    Nop, // should we have one?
```

Comment:
I say why not? It can be useful sometimes for a compiler

Reply:
I would agree, though having said so I have seen that used as padding in instruction stream that need alignment (e.g. word alignment for switch data at the end of the code). 
Having said so I would agree with you and it has some advantages, for instance, it's the easiest instruction to use to test big code and things like that....

Reply:
Yes, please. It seems to be quite important for supporting accurate association of verifier annotations with bytecode offsets.